zerver/actions/message_send.py

```diff
@@ -239,6 +240,8 @@ def should_send(setting: str, row: Dict[str, Any]) -> bool:
             # This implements the structure that the UserProfile stream notification settings
             # are defaults, which can be overridden by the stream-level settings (if those
             # values are not null).
+            if row["user_profile_id"] in user_ids_unmuting_topic:
+                return True
```

Comment:
```diff
         def should_send(setting: str, row: Dict[str, Any]) -> bool:
             # This implements the structure that the UserProfile stream notification settings
             # are defaults, which can be overridden by the stream-level settings (if those
             # values are not null).
             if row["is_muted"]:
+               if row["user_profile_id"] not in user_ids_unmuting_topic:
+                  return False
             if row["user_profile_id"] in user_ids_muting_topic:
                 return False
             if row[setting] is not None:
                 return row[setting]
             return row["user_profile_" + setting]
   ```
   
   This satisfies the same logic.

Reply:
Yes, makes sense 👍

Reply:
Updated `should_send` function:
```diff
        def should_send(setting: str, row: Dict[str, Any]) -> bool:
            # This implements the structure that the UserProfile stream notification settings
            # are defaults, which can be overridden by the stream-level settings (if those
            # values are not null).
-            if row["is_muted"]:
+            if row["is_muted"] and row["user_profile_id"] not in user_ids_unmuting_topic:
                return False
            if row["user_profile_id"] in user_ids_muting_topic:
                return False
            if row[setting] is not None:
                return row[setting]
            return row["user_profile_" + setting]
```

But, as mentioned by `abhijeetbodas2001` 
> I think the logic here is going to get very complicated very quickly when we implement the visibility policies. I had tried to extract out this logic into the notification_data.py file so the reading complexity of the message-send codepath is reduced. @timabbott if you could take a look at https://github.com/zulip/zulip/pull/24196, that would be great!

It needs attention from @timabbott

Reply:
OK took a look at #24196; looks good aside from a small tweak to avoid losing the `wildcard_mention_user_ids` optimization.

Reply:
And I think this should be able to be pretty easily rebased on top of that once it's merged; we just might need to pass a map from `user_id => visibility_policy` instead of `user_ids_muting_topics`, or perhaps just have a `user_ids_unmuting_topics` separate data structure; I'm not sure which is cleaner but the latter might be easier to work with.

Reply:
Rebased + added the required logic to [`user_allows_notifications_in_StreamTopic`](https://github.com/zulip/zulip/pull/24236/commits/15e3edc3e71bc3a9ce63373551be04679be44b8b#diff-8319130d25323802a8a47fa49f43d3c5e4c73567aa1e2fc95adb6d4afa68646aR188)